src/MongoDB.Driver.Core/Core/Compression/Zstandard/ZstandardNativeWrapper.cs

```diff
@@ -0,0 +1,288 @@
+            _compressionMode = compressionMode;
+
+            switch (_compressionMode)
+            {
+                case CompressionMode.Compress:
+                    // calculate recommended size for input buffer
+                    _recommendedZstreamInputSize = Zstandard64NativeMethods.ZSTD_CStreamInSize();
+                    // calculate recommended size for output buffer. Guarantee to successfully flush at least one complete compressed block
+                    _recommendedZstreamOutputSize = Zstandard64NativeMethods.ZSTD_CStreamOutSize();
+                    _zstreamPointer = Zstandard64NativeMethods.ZSTD_createCStream(); // create resource
+                    break;
+                case CompressionMode.Decompress:
+                    // calculate recommended size for input buffer
+                    _recommendedZstreamInputSize = Zstandard64NativeMethods.ZSTD_DStreamInSize();
+                    // calculate recommended size for output buffer. Guarantee to successfully flush at least one complete block in all circumstances
+                    _recommendedZstreamOutputSize = Zstandard64NativeMethods.ZSTD_DStreamOutSize();
+                    _zstreamPointer = Zstandard64NativeMethods.ZSTD_createDStream(); // create resource
+                    break;
+            }
+        }
+
+        public int RecommendedInputSize => (int)_recommendedZstreamInputSize;
+        public int RecommendedOutputSize => (int)_recommendedZstreamOutputSize;
+
+        // public methods
+        public void Compress(
+            OperationContext operationContext,
+            int compressedSize,
+            int uncompressedSize,
+            out int compressedBufferPosition,
+            out int uncompressedBufferPosition)
+        {
+            Ensure.IsNotNull(operationContext, nameof(operationContext));
+            Ensure.That(_compressionMode == CompressionMode.Compress, nameof(_compressionMode));
+
+            InitializeIfNotAlreadyInitialized();
+
+            // compressed data
+            var outputNativeBuffer = CreateNativeBuffer(
+                operationContext.CompressedPinnedBufferWalker, // operation result
+                (ulong)compressedSize);
+
+            // uncompressed data
+            var inputNativeBuffer = CreateNativeBuffer(
+                operationContext.UncompressedPinnedBufferWalker,
+                (ulong)uncompressedSize);
+
+            // compress inputNativeBuffer to outputNativeBuffer
+            Zstandard64NativeMethods.ZSTD_compressStream(
+                _zstreamPointer,
+                outputNativeBuffer,
+                inputNativeBuffer);
+
+            compressedBufferPosition = (int)outputNativeBuffer.Position;
+
+            // calculate progress in inputNativeBuffer
+            uncompressedBufferPosition = (int)inputNativeBuffer.Position;
+            operationContext.UncompressedPinnedBufferWalker.Offset += uncompressedBufferPosition;
```

Comment:
I think we can all agree that adding two positions together is wrong.

I think what you are saying is that the uncompressedBufferPosition is not really a position. You are claiming it is a count. 

If that's really true rename the variable.

I'm also willing to simply state that I think it is wrong but that I am overruled by you and other reviewers.